Enhance and update documentation (#351)
kubernetes-sigs/kueue: docs/concepts/cluster_queue.md

# Cluster Queue

A ClusterQueue is a cluster-scoped object that governs a pool of resources
such as CPU, memory and hardware accelerators. A `ClusterQueue` defines:

> - The [resource _flavors_](#resourceflavor-object) that it manages, with usage

Reviewer: ```suggestion
- The [resource _flavors_](#resourceflavor-object) that the cluster manages, with usage
```

I'm assuming "it" refers to "the cluster" but please edit accordingly, but we must mention what "it" means.

You can specify the quota as a [quantity](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/quantity/).

## Resources

In a ClusterQueue, you can define quotas for multiple [compute resources](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#resource-types)

> (cpu, memory, GPUs, etc.).

Reviewer: ```suggestion
(CPU, memory, GPUs, etc.).
```

> For each resource, you can define quotas for multiple _flavors_. A
> flavor represents different variations of a resource. The variations can be
> defined in a [ResourceFlavor object](#resourceflavor-object).

Reviewer 2: ```suggestion
For each resource, you can define quotas for multiple _flavors_. Flavors represent different variations of a resource (for example, different GPU models). A flavor can be defined using a [ResourceFlavor object](#resourceflavor-object).
```

In a process called [admission](.#admission), Kueue assigns

> [Workload pod sets](workload.md#pod-sets) a flavor for each resource it requests.

Reviewer: ```suggestion
to the [Workload pod sets](workload.md#pod-sets) a flavor for each resource Kueue requests.
```
Not sure if "it" refers to "Kueue" but we need to replace "it" with the object name.

Kueue assigns the first flavor in the ClusterQueue's `.spec.resources[*].flavors`
list that has enough unused `min` quota in the ClusterQueue or the
ClusterQueue's [cohort](#cohort).

### Codepedent resources

It is possible that multiple resources are tied to the same flavors. This is
typical for `cpu` and `memory`, where the flavors are generally tied to a

> machine family or availability guarantees.

Reviewer: ```suggestion
machine family or availability policies. When two or more resources match their flavors,
they are said to be codependent resources. 
```

> machine family or availability guarantees.

Reviewer 2: ```suggestion
machine family or provisioning model (e.g., spot vs standard).
```

> If this is the case, the resources in the ClusterQueue must list the same
> flavors in the same order. When two or more resources match their flavors,
> they are said to be codependent. During admission, for each pod set in a
> Workload, Kueue assigns the same flavor to the codependent resources that the
> pod set requests.

Reviewer: ```suggestion
To manage codependent resources, list the resources in the ClusterQueue in the same
flavors order. During admission, for each pod set in a
Workload, Kueue assigns the same flavor to the codependent resources that the
pod set requests.
```

- architecture (ex: x86 vs ARM CPUs)
- brands and models (ex: Radeon 7000 vs Nvidia A100 vs T4 GPUs)

> A ResourceFlavor is an object that represents these variations and allows you

Reviewer: ```suggestion
A ResourceFlavor is an object that represents these resource variations and allows you
```

> A ResourceFlavor is an object that represents these variations and allows you
> to associate them with node labels and taints.

Reviewer 2: ```suggestion
A ResourceFlavor is an object that represents these variations and allows you
to associate them with the labels and taints of the nodes that offer those resource flavors.
```

   didn't specify them already.

   For example, for a [batch/v1.Job](https://kubernetes.io/docs/concepts/workloads/controllers/job/),

>    Kueue adds the labels to the `.spec.template.spec.nodeSelector` field. This
>    guarantees that the workload Pods run on the nodes associated to the flavor
>    that Kueue decided that the workload should use.

Reviewer 2: ```suggestion
   Kueue adds the labels to the `.spec.template.spec.nodeSelector` field. This
   guarantees that the workload Pods can only be scheduled on the nodes targeted by the flavor
   that Kueue assigned to the workload.
```

Author: nice, I was having a hard time with this one :)

Taints on the ResourceFlavor work similarly to [node taints](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/).
For Kueue to admit a workload to use the ResourceFlavor, the PodSpecs in the
workload should have a toleration for it. As opposed to the behavior for

> [ResourceFlavor labels](#resourceflavor-labels), Kueue will not add tolerations

Reviewer: ```suggestion
[ResourceFlavor labels](#resourceflavor-labels), Kueue does not add tolerations
```

### Flavors and borrowing semantics

When borrowing, Kueue satisfies the following admission semantics:

> - When assigning flavors, Kueue goes through the list of flavors in the
>   ClusterQueue's `.spec.resources[*].flavors`. For each flavor, Kueue attempts
>   to fit a Workload's pod set using the `min` quota of the ClusterQueue or the
>   unused `min` quota of other ClusterQueues in the cohort, up to the `max` quota
>   of the ClusterQueue. If the workload doesn't fit, Kueue proceeds evaluating the next
>   flavor in the list.

Reviewer: ```suggestion
- When assigning flavors, Kueue goes through the list of flavors in the
  ClusterQueue's `.spec.resources[*].flavors`. For each flavor, Kueue attempts
  to fit a Workload's pod set using the `min` quota of the ClusterQueue or the
  unused `min` quota of other ClusterQueues in the cohort. If the workload doesn't fit, Kueue proceeds evaluating the next flavor in the list. Kueue attempts to fit workloads until the `max` quota
  of the ClusterQueue is reached. 
```

> - A ClusterQueue can only borrow quota of flavors it defines and it can only
>   borrow quota for one flavor.

Reviewer: ```suggestion
- A ClusterQueue can only borrow quota of flavors that ClusterQueue defines. ClusterQueue can only
  borrow quota for one flavor.
```

>   borrow quota for one flavor.

Reviewer 2: I wouldn't mention "and it can only borrow quota for one flavor" because that also applies in the no borrowing case as well.
